[PVR] Echo up important Status info to Timer Rules

metaron-uk opened the pull request:
- Number of 'Will Record' Child Timers
- Presence of InProg Child Timers
- Presence of Conflict_NOK or Error Child Timers

Also adds 'Completed' status as distinct from 'Enabled'
![screenshot037](https://cloud.githubusercontent.com/assets/12870817/12071357/93f10002-b09f-11e5-93b2-eef4388a54db.png)

No sure how this will play with all existing pvr clients, but I can't see it breaking anything.
Small tweaks may need to be made to the reported status of 'Timer Rules' in each client to get it to work fully as intended. I'm hoping the logic is sound so it should just 'work'.

If a client already echos up 'Conflict / Error / Recording' status the behaviour should be unchanged. 
Only a status of 'NEW' or 'SCHEDULED' and in the case of recording in prog, 'COMPLETED' will be replaced with the new information based on child timer states.
(I made some changes to pvr.mythtv to get 'COMPLETED' to work as shown which I will PR if this is accepted)

I'm a little uncomfortable about putting the child timer status reset in CPVRTimerInfoTag::UpdateEntry and the updates in CPVRTimers::UpdateEntries. Happy to move it all to CPVRTimers::UpdateEntries if there are strong opinions.

pings: @ksooo, @ryangribble, @janbar, @jalle19

metaron-uk committed:
[PVR] Echo up important Status info to Timer Rules

 - Number of Scheduled Child Timers (including New/Conflict_OK)
 - Presence of Child Timers currently recording
 - Presence of Conflict_NOK or Error Child Timers

Also adds 'Completed' status as distinct from 'Enabled'

xhaggi commented on xbmc/pvr/timers/PVRTimers.cpp:
```diff
@@ -248,6 +248,47 @@ bool CPVRTimers::UpdateEntries(const CPVRTimers &timers)
+          case PVR_TIMER_STATE_NEW:
+          case PVR_TIMER_STATE_SCHEDULED:
+          case PVR_TIMER_STATE_CONFLICT_OK:
+            parentTimer->m_iActiveChildTimers ++;
```
Please remove the space before `++` here and below

metaron-uk replied:
> Please remove the space before ++ here and below

Done - separate commit below. To be squashed once discussions are complete.

Glenn-1990 commented:
@metaron-uk First of all nice work.

I'm not sure about the "completed" status, in my eyes a completed timer is no timer anymore, it's a recording. You seems to depend on PVR_TIMER_STATE_COMPLETED for the completed state, but do clients use that state for repeating schedules? At least tvh will only use PVR_TIMER_STATE_SCHEDULED or PVR_TIMER_STATE_DISABLED for repeating schedules as "completed" is a bit misleading here.

@ksooo Maybe the api should define which timer states are for parents and which are for childs?

Jalle19 commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -9491,7 +9491,15 @@ msgctxt "#19241"
-#empty strings from id 19242 to 19243
+#: xbmc/pvr/timers/PVRTimerInfoTag.cpp
+msgctxt "#19242"
+msgid "%d will record"
```
I suggest using "upcoming" instead of "will record"

metaron-uk replied:
@Jalle19 @xhaggi - Problems with 'upcoming'
1) 1 upcoming  / 2,3,4,etc... upcomings - i.e. two different strings needed. Not necessarily a problem but...
2) You can set pvr.mythtv to include 'inactive' matches for each rule (repeats, duplicates etc...) in the timers list. This PR logic currently excludes child timers with 'won't record' states (conflict_nok etc..).  I think 'will record' more clearly explains why these 'inactive' child timers aren't included in the count.
I'm open to further discussion / suggestion but still think with the current PRs logic 'will record' is better.

xhaggi commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -9491,7 +9491,15 @@ msgctxt "#19241"
+#: xbmc/pvr/timers/PVRTimerInfoTag.cpp
+msgctxt "#19242"
+msgid "%d will record"
+msgstr ""
```
agree

metaron-uk commented:
My idea with using the 'Completed' status is to allow the client to indicate (if it has the information) that a rule has actually carried out a recording successfully (see https://github.com/metaron-uk/xbmc/blob/sorting/xbmc/addons/include/xbmc_pvr_types.h#L180).

Mythv includes 'will next record' and 'last recorded' timestamps for each rule updated by the scheduler. When the scheduler finds a match for the rule, it updates the 'will next record' timestamp with the first match. When a recording starts, the 'last recorded' timestamp is updated.

By setting rules with a 'last recorded' timestamp but no 'will next record' timestamp to 'COMPLETED', rules for which a recording was performed, state 'Completed', rules which are about to do something say 'x will record' and those which haven't done anything yet say 'Enabled'.

This way you can easily spot rules which have already done something and aren't planning to do anything else within available guide data and remove/edit them if you want.
Rules which are still waiting for a match (e.g. because the latest starwars hasn't been shown on TV yet) show 'Enabled' until they have done something.

If your backend cleans up after itself and deletes 'Completed' rules automatically, you'll never see this status, but if not, it helps separate 'finished' rules from 'not yet started' ones.

ksooo commented:
> By setting rules with a 'last recorded' timestamp but no 'will next record' timestamp to 'COMPLETED'

 Sorry, but imo using "Completed" for timer rules does not fit. As you explain it it is more "Currently nothing scheduled". "Completed" for me means that there will be never ever new timers scheduled by this rule which could only be the case if you set an end day and time for the rule.

metaron-uk commented:
>>  By setting rules with a 'last recorded' timestamp but no 'will next record' timestamp to 'COMPLETED'

> Sorry, but imo using "Completed" for timer rules does not fit. As you explain it it is more "Currently nothing scheduled". "Completed" for me means that there will be never ever new timers scheduled by this rule which could only be the case if you set an end day and time for the rule.

What about a 'Record one' rule? pvr.mythtv has these, so assuming 'one' has been recorded successfully, the rule is 'completed' without needing an end date. The definition in the API "/*!< @brief the recording completed successfully */" seems fairly clear and seems to be compatible with displaying 'Completed' in this scenario.

I agree the logic I've described for setting 'Completed' status for a rule is less than perfect (especially for 'record all' rules), but IMO how best to decide if a rule is 'complete' is something for the client/backend.

The other states seem to be:
* 'NEW' "/*!< @brief the timer was just created on the backend and is not yet active. This state must not be used for timers just created on the client side. */" and 
* 'SCHEDULED' "/*!< @brief the timer is scheduled for recording */"

New covers 'hasn't resulted in any intended recordings yet' and Scheduled covers 'is currently resulting in intended recordings'.

There isn't a 'has caused recordings before, might do so again, but isn't currently planning to do so' status. The choice would appear to be 'New' , 'Completed', or changing the API to add 'Resting'.

With the logic in this PR, if the client sets 'NEW' or 'SCHEDULED without any child timers the display will show 'Enabled'. If it sets 'COMPLETED' the display shows 'Completed' unless it has children currently recording. 

I agree allowing in-prog child recordings to over-ride 'COMPLETED' status might be considered superfluous, but it does make the code in pvr.mythv for 'Completed' easier :wink:. I can take this bit out if you like.

I'd be happy to raise a separate PR to add ```PVR_TIMER_STATE_RESTING = 10 /*!< @brief the timer rule is active, has previously recorded, but currently can't find a match in available guide data */``` if you think this state is worth separating out from 'New' or 'Completed'.

metaron-uk commented:
Rebased and squashed (the 'spare' strings were used by another merged PR).

I agree my criteria for setting 'Completed' rule status was flawed (I'm pondering more valid alternatives), but I still think 'Completed' is valid for either a rule (i.e. for completed manual / once rules) or an individual timer to report. I believe if the client reports it, displaying it would be helpful and informative.

I hope my '%d will record' explanation was good enough to justify not re-naming it '%d upcoming(s)'

@ksooo / @xhaggi / @Glenn-1990 / others - have I convinced you, or do you have use cases which won't work with these names which I can ponder?

Any opinions on adding a new 'RESTING' timer state?

ksooo commented:
> but I still think 'Completed' is valid for either a rule (i.e. for completed manual / once rules)

 Agree to "record one" rule.
 Do not understand what a "completed manual rule" is.

> Any opinions on adding a new 'RESTING' timer state (in a separate PR)?
> There isn't a 'has caused recordings before, might do so again, but isn't currently planning to do so' status.

 I do not see the need for this new state. "RESTING" can easily be calculated by Kodi from existing information: state == SCHEDULED && rule_has_no_children. No need for another PVR API timer state. What you want to achieve imo is just about user interface, not PVR API. UI representation of the kodi-calculated "state" could be something like "resting", but again, no need to extend API for this.

metaron-uk commented:
> Do not understand what a "completed manual rule" is.

In mythtv all 'upcoming' timers have a parent rule on the backend, even manual or 'this showing' ones. Currently pvr.mythtv client doesn't transfer oneshot 'manual' rules to kodi core for consideration unless it has an associated 'upcoming', but they could potentially still accumulate on the backend as 'completed manual rules' (I'm not 100% sure about this as I don't seem to have any languishing on my backend, maybe it deletes them). The mythtv client could be extended to transfer these rules to kodi core for housekeeping purposes. Can't get too excited about this though.

> I do not see the need for this new state. "RESTING"...

OK by me, one less API change PR :-) 

Let me know if anything else occurs to you.

ksooo commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -9553,7 +9553,15 @@ msgctxt "#19254"
 msgstr ""
 
-#empty strings from id 19255 to 19256
+#: xbmc/pvr/timers/PVRTimerInfoTag.cpp
```
Please add a short explanation for the purpose of this string. <code>#. blah blah</code>

ksooo commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -9553,7 +9553,15 @@ msgctxt "#19254"
+msgid "%d will record"
+msgstr ""
+
+#: xbmc/pvr/timers/PVRTimerInfoTag.cpp
```
Same here.

ksooo replied:
What do you thing about "%d scheduled"? You never know whether those actually **will** be recorded. ;-)

metaron-uk replied:
I agree "%d scheduled" it is.

ksooo commented on xbmc/pvr/timers/PVRTimers.cpp:
```diff
@@ -248,6 +248,47 @@ bool CPVRTimers::UpdateEntries(const CPVRTimers &timers)
   }
 
+  /* go through the timer list and check for parent timers */
+  for (MapTags::const_iterator it = timers.m_tags.begin(); it != timers.m_tags.end(); ++it)
```
c++11: <code>for (const auto &tagsEntry : m_tags)</code> - much more elegant, imo.

ksooo commented on xbmc/pvr/timers/PVRTimers.cpp:
```diff
@@ -248,6 +248,47 @@ bool CPVRTimers::UpdateEntries(const CPVRTimers &timers)
+  /* go through the timer list and check for parent timers */
+  for (MapTags::const_iterator it = timers.m_tags.begin(); it != timers.m_tags.end(); ++it)
+  {
+    for (VecTimerInfoTag::const_iterator timerIt = it->second->begin(); timerIt != it->second->end(); ++timerIt)
```
c++11:  <code>for (const auto &timersEntry : *tagsEntry.second)</code> - much more elegant, imo.

metaron-uk replied:
You're right it is. Thanks for another informative c++ lesson!

ksooo commented on xbmc/pvr/timers/PVRTimers.cpp:
```diff
@@ -248,6 +248,47 @@ bool CPVRTimers::UpdateEntries(const CPVRTimers &timers)
+          case PVR_TIMER_STATE_ABORTED:
+          case PVR_TIMER_STATE_CANCELLED:
+          case PVR_TIMER_STATE_DISABLED:
+            break;
```
I would add some comment like "nothing to do for these states" to indicate that doing nothing here is not a mistake / forgotten to implement...

metaron-uk replied:
comments added here and in strings.po

ksooo commented on xbmc/pvr/timers/PVRTimers.cpp:
```diff
@@ -248,6 +248,47 @@ bool CPVRTimers::UpdateEntries(const CPVRTimers &timers)
+    for (VecTimerInfoTag::const_iterator timerIt = it->second->begin(); timerIt != it->second->end(); ++timerIt)
+    {
+      /* check if the parent timer is present */
+      CPVRTimerInfoTagPtr parentTimer = GetByClient((*timerIt)->m_iClientId, (*timerIt)->m_iParentClientIndex);
```
Please add
<pre>
unsigned int iRuleId = timersEntry->GetTimerRuleId();
if (iRuleId != PVR_TIMER_NO_PARENT)
</pre>

before doing the expensive call to <code>GetByClientId</code> (look at the implementation to see why it is expensive!)

And, "const as const can" to help the compiler to generate efficient code: <code>const CPVRTimerInfoTagPtr</code>, please.

And even more better <code>const CPVRTimerInfoTagPtr parentTimer(GetByClient(timersEntry->m_iClientId, timersEntry->GetTimerRuleId()));</code>. Always prefer copy ctor (1 call) over standard ctor + assignment operator (2 calls). Even though compilers could optimize this, you never know they actually will do. If you write it like this, compiler will do. :-)

metaron-uk replied:
Have tweaked in line with your suggestions - seems to read nicely and is also more efficient.

ksooo commented on xbmc/pvr/timers/PVRTimers.cpp:
```diff
@@ -248,6 +248,47 @@ bool CPVRTimers::UpdateEntries(const CPVRTimers &timers)
+      CPVRTimerInfoTagPtr parentTimer = GetByClient((*timerIt)->m_iClientId, (*timerIt)->m_iParentClientIndex);
+      if (parentTimer)
+      {
+        CPVRTimerInfoTagPtr childTimer = GetByClient((*timerIt)->m_iClientId, (*timerIt)->m_iClientIndex);
```
const plus copy ctor

ksooo replied:
This line makes not much sense, imo. <code>\*timerIt</code> (or <code>timersEntry</code> after c++11 rewrite) IS the childtimer. So, the line above should always return <code>*timerIt</code> / <code>timersEntry</code>!

ksooo commented on xbmc/pvr/timers/PVRTimers.cpp:
```diff
@@ -248,6 +248,47 @@ bool CPVRTimers::UpdateEntries(const CPVRTimers &timers)
+          case PVR_TIMER_STATE_NEW:
+          case PVR_TIMER_STATE_SCHEDULED:
+          case PVR_TIMER_STATE_CONFLICT_OK:
+            parentTimer->m_iActiveChildTimers++;
```
Argh! Public members are really bad style (don't do it like people did bad before you)! Please add a setter method to encapsulate the data.

ksooo replied:
I would do this very different, btw. Add a new method <code>bool CPVRTimerInfoTag::UpdateChildStatus(const CPVRTimerInfoTagPtr &child)</code> and move the code that currently sets the new members to it. This way, you will not need the setters for the members as they will not be accessed from outside the class. Anyway, make the new members private, please. 

Then, here would stay just:

<pre>
if (childTimer)
 parentTimer->UpdateChildState(childTimer);
</pre>

ksooo commented on xbmc/pvr/timers/PVRTimerInfoTag.cpp:
```diff
@@ -612,6 +640,12 @@ bool CPVRTimerInfoTag::UpdateEntry(const CPVRTimerInfoTagPtr &tag)
+  m_bHasChildConflictNOK = false;
+  m_bHasChildRecording = false;
+  m_bHasChildErrors = false;
+  // Updated in CPVRTimers::UpdateEntries which should run after this...
```
This comment is a clear sign of bad design. Separation of concerns! Do it like I suggested (the new CPVRTimerInfoTag method) and comments like this can go away. ;-)

metaron-uk replied:
I agree.
I've kept a comment referring to UpdateChildState as well as adding a description for both fucntions and placing them next to eachother in PVRTimerInfoTag.h.
If you have any better ideas, I'm sure you will let me know.

metaron-uk commented:
Comments incorporated. Let me know if I've missed anything.

metaron-uk commented:
OK now?

metaron-uk commented:
Fingers crossed! Many thanks for the help and and the constructive criticism @ksooo.

ksooo commented:
You're welcome and thanks for this nice improvement.

ksooo commented:
jenkins build this please

ksooo commented:
WIN-32 jenkins error is unrelated.

ksooo commented:
Runtime-tested. Works like expected.
